torch/csrc/api/include/torch/detail/ordered_dict.h

```diff
@@ -0,0 +1,204 @@
+#pragma once
+
+#include <ATen/Error.h>
+
+#include <cstdint>
+#include <functional>
+#include <initializer_list>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace detail {
+
+/// A simple ordered dictionary implementation, akin to Python's `OrderedDict`.
+template <typename T>
+class OrderedDict {
```

Comment:
String keys only. Are we sure we want to call it `OrderedDict`? Python's ordered dict supports other types of keys.

Reply:
I think it's better to start out specific like this and then make it more general when we need it

Reply:
I'm fine with it being string keys only, but then don't name it OrderedDict?

Reply:
I've added a `Key` template parameter.